common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+ *                                                 be expected to finish.
+ * @property   {Number}     correlationId          The put token operation was sent with a message id.  The response
+ *                                                 to the put token operation will contain this message id as the
+ *                                                 correlation id.  This id is a uuid.
+ *
+ */
+interface PutTokenOperation {
+  putTokenCallback: (err: Error | null, result?: any) => void;
+  expirationTime: number;
+  correlationId: string;
+}
+
+/**
+ * @class      module:azure-iot-amqp-base.PutTokenStatus
+ * @classdesc  Describes the state of the "put token" feature of the client that enables Claim-Based-Security authentication.
+ *
+ * @property   {Function}   outstandingPutTokens                This array will hold outstanding put token operations.  The array has
+ *                                                              a monotonically increasing time ordering.  This is effected by the nature
+ *                                                              of inserting new elements at the end of the array.  Note that elements may
+ *                                                              be removed from the array at any index.
+ * @property   {Number}     numberOfSecondsToTimeout            Currently a fixed value.  Could have a set option if we want to make this configurable.
+ * @property   {Number}     putTokenTimeOutExaminationInterval  While there are ANY put token operations outstanding a timer will be invoked every
+ *                                                              10 seconds to examine the outstandingPutTokens array for any put tokens that may have
+ *                                                              expired.
+ * @property   {Number}     timeoutTimer                        Timer used to trigger examination of the outstandingPutTokens array.
+ */
+class PutTokenStatus {
+    outstandingPutTokens: PutTokenOperation[] = [];
+    numberOfSecondsToTimeout: number = 120;
+    putTokenTimeOutExaminationInterval: number = 10000;
+    timeoutTimer: number;
+}
+
+export class ClaimsBasedSecurityAgent {
+  private static _putTokenSendingEndpoint: string = '$cbs';
+  private static _putTokenReceivingEndpoint: string = '$cbs';
+  private _amqp10Client: amqp10.AmqpClient;
+  private _errorHandler: (err: Error)  => void;
+  private _fsm: machina.Fsm;
+  private _senderLink: SenderLink;
+  private _receiverLink: ReceiverLink;
+  private _putToken: PutTokenStatus = new PutTokenStatus();
+  private _putTokenQueue: {
+    audience: string,
+    token: string,
+    callback: (err?: Error) => void
+  }[];
+
+  constructor(amqp10Client: amqp10.AmqpClient) {
+    this._errorHandler = (err: Error): void => {
+      this._fsm.handle('detach', undefined, err);
+    };
+
+    this._amqp10Client = amqp10Client;
```

Comment:
>amqp10Client [](start = 25, length = 12)

add falsy check?

Reply:
since this is an internal API the compiler does that for us :)

---
In reply to: [139536227](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139536227) [](ancestors = 139536227)